plugin/cache/handler.go

```diff
@@ -38,9 +38,13 @@ func (c *Cache) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg)
 	if i != nil {
 		ttl = i.ttl(now)
 	}
-	if i == nil {
+	if i == nil || (ttl < 0 && c.staleFetchBefore) {
+		// Try to resolve upstream if we have no cached record or the cached one is stale and staleFetchBefore is set to true.
 		crr := &ResponseWriter{ResponseWriter: w, Cache: c, state: state, server: server, do: do}
-		return c.doRefresh(ctx, state, crr)
+		rcode, err := c.doRefresh(ctx, state, crr)
+		if i == nil || rcode == dns.RcodeSuccess {
```

Comment:
Does this intend to ignore any responses that are not "success" and serve stale cache _instead_?
I don't think we can do that.  For example, NXDOMAIN is a valid response, it means the record is currently (and authoritatively) non-existent and that should be returned to the client.  Also, if the response from "upstream" _was_ NXDOMAIN, then this would end up sending two separate responses to the client, the NXDOMAIN response, followed by the stale response (which would probably be ignored by the client).

I think the only reason to serve stale cache in the always-check-before-sending-stale mode, would be is if the upstream server does not respond (i.e. we don't know what the state of the upstream record is).  You might need to wrap the response writer with one that remembers/exposes if a message was written to the client.

Reply:
The relevant section of the RFC... https://datatracker.ietf.org/doc/html/rfc8767#section-4

>Answers from authoritative servers that have a DNS response code of
   either 0 (NoError) or 3 (NXDomain) and the Authoritative Answer (AA)
   bit set MUST be considered to have refreshed the data at the
   resolver.  Answers from authoritative servers that have any other
   response code SHOULD be considered a failure to refresh the data and
   therefore leave any previous state intact.  See Section 6 for a
   discussion.

Reply:
IOW - per the RFC - NOERROR and NXDOMAIN responses MUST get sent to the client, and the stale cache is replaced. For other response codes, e.g. SERVFAIL/REFUSED, the response SHOULD be ignored if stale cache is present, and stale cache sent to client. Section 6 of the RFC discusses rationale for treating various responses codes differently (in effect - a case for making the behavior configurable for response codes other than NOERROR/NXDOMAIN).